File: 90 DWT 3/MeasurementsReportViewController.m
//
//  MeasurementsReportViewController.m
//  90 DWT 1
//
//  Created by Jared Grant on 7/14/12.
//  Copyright (c) 2012 g-rantsoftware.com. All rights reserved.
//

#import "MeasurementsReportViewController.h"

@interface MeasurementsReportViewController ()

@end

@implementation MeasurementsReportViewController

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        // Custom initialization
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
	// Do any additional setup after loading the view.
    UIColor *lightGrey = [UIColor colorWithRed:234/255.0f green:234/255.0f blue:234/255.0f alpha:1.0f];
    self.view.backgroundColor = lightGrey;
    
    [self loadDictionary];
    [self.htmlView loadHTMLString:[self createHTML] baseURL:nil];
    self.htmlView.backgroundColor = [UIColor clearColor];
    self.htmlView.opaque = NO;
}

- (BOOL)shouldAutorotateToInterfaceOrientation:(UIInterfaceOrientation)interfaceOrientation
{
    return (interfaceOrientation == UIInterfaceOrientationPortrait);
}

- (void)emailSummary {
    // Send email
    
    // Create an array of measurements to iterate thru when building the table rows.
    NSArray *measurementsArray = @[self.month1Dict, self.month2Dict, self.month3Dict, self.finalDict];
    NSArray *measurementsMonth = @[@"Start Month 1", @"Start Month 2", @"Start Month 3", @"Final"];
    
    NSMutableString *writeString = [NSMutableString stringWithCapacity:0];
    [writeString appendString:[NSString stringWithFormat:@"Month,Weight,Chest,Left Arm,Right Arm,Waist,Hips,Left Thigh,Right Thigh\n"]];
    
    for (int i = 0; i < measurementsMonth.count; i++) {
        [writeString appendString:[NSString stringWithFormat:@"%@,%@,%@,%@,%@,%@,%@,%@,%@\n",
                                   measurementsMonth[i],
                                   measurementsArray[i][@"Weight"],
                                   measurementsArray[i][@"Chest"], 
                                   measurementsArray[i][@"Left Arm"], 
                                   measurementsArray[i][@"Right Arm"], 
                                   measurementsArray[i][@"Waist"], 
                                   measurementsArray[i][@"Hips"],
                                   measurementsArray[i][@"Left Thigh"], 
                                   measurementsArray[i][@"Right Thigh"]]];
    }
    
    NSData *csvData = [writeString dataUsingEncoding:NSASCIIStringEncoding];
    NSString *fileName = [self.navigationItem.title stringByAppendingString:@" Measurements.csv"];
    
    // Create MailComposerViewController object.
    MFMailComposeViewController *mailComposer;
    mailComposer = [[MFMailComposeViewController alloc] init];
    mailComposer.mailComposeDelegate = self;
    mailComposer.navigationBar.tintColor = [UIColor whiteColor];
    
    // Array to store the default email address.
    NSArray *emailAddresses; 
    
    // Get path to documents directory to get default email address.
    NSString *docDir = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES)[0];
    NSString *defaultEmailFile = nil;
    defaultEmailFile = [docDir stringByAppendingPathComponent:@"Default Email.out"];
    
    if ([[NSFileManager defaultManager] fileExistsAtPath:defaultEmailFile]) {
        NSFileHandle *fileHandle = [NSFileHandle fileHandleForReadingAtPath:defaultEmailFile];
        
        NSString *defaultEmail = [[NSString alloc] initWithData:[fileHandle availableData] encoding:NSUTF8StringEncoding];
        [fileHandle closeFile];
        
        // There is a default email address.
        emailAddresses = @[defaultEmail];
    }
    else {
        // There is NOT a default email address.  Put an empty email address in the arrary.
        emailAddresses = @[@""];
    }
    
    [mailComposer setToRecipients:emailAddresses];
    
    NSString *subject = @"90 DWT 1";
    subject = [subject stringByAppendingFormat:@" %@ Measurements", self.navigationItem.title];
    [mailComposer setSubject:subject];
    [mailComposer addAttachmentData:csvData mimeType:@"text/csv" fileName:fileName];
    [self presentViewController:mailComposer animated:YES completion:^{
        [[UIApplication sharedApplication] setStatusBarStyle:UIStatusBarStyleLightContent];
    }];
}

- (IBAction)actionSheet:(id)sender {
    UIActionSheet *action = [[UIActionSheet alloc] initWithTitle:@"Share" delegate:self cancelButtonTitle:@"Cancel" destructiveButtonTitle:nil otherButtonTitles:@"Email", @"Facebook", @"Twitter", nil];
    [action showInView:self.view];
}

- (void)actionSheet:(UIActionSheet *)actionSheet clickedButtonAtIndex:(NSInteger)buttonIndex {
    
    if (buttonIndex == 0) {
        [self emailSummary];
    }
    
    if (buttonIndex == 1) {
        [self facebook];
    }
    
    if (buttonIndex == 2) {
        [self twitter];
    }
}

- (void)mailComposeController:(MFMailComposeViewController *)controller didFinishWithResult:(MFMailComposeResult)result error:(NSError *)error
{
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (void)loadDictionary {
    // Get path to documents directory
    NSString *docDir = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES)[0];
    NSString *dictionaryFile = nil;
    
    // Month 1
    dictionaryFile = [docDir stringByAppendingPathComponent:@"Start Month 1 Measurements.out"];
    if ([[NSFileManager defaultManager] fileExistsAtPath:dictionaryFile]) {
        self.month1Dict = [NSDictionary dictionaryWithContentsOfFile:dictionaryFile];
    }
    else {
        self.month1Dict = @{@"Weight": @"0",
                      @"Chest": @"0",
                      @"Left Arm": @"0",
                      @"Right Arm": @"0",
                      @"Waist": @"0",
                      @"Hips": @"0",
                      @"Left Thigh": @"0",
                      @"Right Thigh": @"0"};
    }
    
    // Month 2
    dictionaryFile = [docDir stringByAppendingPathComponent:@"Start Month 2 Measurements.out"];
    if ([[NSFileManager defaultManager] fileExistsAtPath:dictionaryFile]) {
        self.month2Dict = [NSDictionary dictionaryWithContentsOfFile:dictionaryFile];
    }
    else {
        self.month2Dict = @{@"Weight": @"0",
                      @"Chest": @"0",
                      @"Left Arm": @"0",
                      @"Right Arm": @"0",
                      @"Waist": @"0",
                      @"Hips": @"0",
                      @"Left Thigh": @"0",
                      @"Right Thigh": @"0"};
    }
    
    // Month 3
    dictionaryFile = [docDir stringByAppendingPathComponent:@"Start Month 3 Measurements.out"];
    if ([[NSFileManager defaultManager] fileExistsAtPath:dictionaryFile]) {
        self.month3Dict = [NSDictionary dictionaryWithContentsOfFile:dictionaryFile];
    }
    else {
        self.month3Dict = @{@"Weight": @"0",
                      @"Chest": @"0",
                      @"Left Arm": @"0",
                      @"Right Arm": @"0",
                      @"Waist": @"0",
                      @"Hips": @"0",
                      @"Left Thigh": @"0",
                      @"Right Thigh": @"0"};
    }
    
    // Final
    dictionaryFile = [docDir stringByAppendingPathComponent:@"Final Measurements.out"];
    if ([[NSFileManager defaultManager] fileExistsAtPath:dictionaryFile]) {
        self.finalDict = [NSDictionary dictionaryWithContentsOfFile:dictionaryFile];
    }
    else {
        self.finalDict = @{@"Weight": @"0",
                     @"Chest": @"0",
                     @"Left Arm": @"0",
                     @"Right Arm": @"0",
                     @"Waist": @"0",
                     @"Hips": @"0",
                     @"Left Thigh": @"0",
                     @"Right Thigh": @"0"};    }
}

- (NSString*)createHTML {
    // Create an array of measurements to iterate thru when building the table rows.
    NSArray *measurementsArray = @[self.month1Dict, self.month2Dict, self.month3Dict, self.finalDict];
    NSArray *measurementsNameArray = @[@"Weight", @"Chest", @"Left Arm", @"Right Arm", @"Waist", @"Hips", @"Left Thigh", @"Right Thigh"];
    
    NSString *myHTML = @"<html><head>";
    
    // Table Style
    myHTML = [myHTML stringByAppendingFormat:@"<STYLE TYPE='text/css'><!--TD{font-family: Arial; font-size: 12pt;}TH{font-family: Arial; font-size: 14pt;}---></STYLE></head><body><table border='1' bordercolor='#3399FF' style='background-color:#CCCCCC' width='%f' cellpadding='2' cellspacing='1'>", (self.htmlView.frame.size.width - 15)];
    
    // Table Headers
    myHTML = [myHTML stringByAppendingString:@"<tr><th style='background-color:#999999'></th><th style='background-color:#999999'>1</th><th style='background-color:#999999'>2</th><th style='background-color:#999999'>3</th><th style='background-color:#999999'>Final</th></tr>"];
    
    // Table Data
    for (int i = 0; i < measurementsNameArray.count; i++) {
        myHTML = [myHTML stringByAppendingFormat:@"<tr><td style='background-color:#999999'>%@</td>", measurementsNameArray[i]];
        
        for (int a = 0; a < measurementsArray.count; a++) {
            myHTML = [myHTML stringByAppendingFormat:@"<td>%@</td>",
                      measurementsArray[a][measurementsNameArray[i]]];
        }
        
        myHTML = [myHTML stringByAppendingString:@"</tr>"];
    }
    
    // HTML closing tags
    myHTML = [myHTML stringByAppendingString:@"</table></body></html>"];
    
    return myHTML;
}
@end

File: 90 DWT 3/MeasurementsReportViewController.swift
//
//  MeasurementsReportViewController.swift
//  90 DWT 3
//
//  Created by Grant, Jared on 8/2/16.
//  Copyright © 2016 Grant, Jared. All rights reserved.
//

import UIKit
import CoreData
import Social

class MeasurementsReportViewController: UIViewController, MFMailComposeViewControllerDelegate, UIPopoverPresentationControllerDelegate, UIPopoverControllerDelegate {
    
    // **********
    let debug = 0
    // **********
    
    @IBOutlet weak var htmlView: UIWebView!
    @IBOutlet weak var shareButton: UIBarButtonItem!
    
    var session = ""
    var month1Array = NSMutableArray()
    var month2Array = NSMutableArray()
    var month3Array = NSMutableArray()
    var finalArray = NSMutableArray()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.loadMeasurements()
        self.htmlView.loadHTMLString(self.createHTML() as String, baseURL: nil)
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
        // Force fetch when notified of significant data changes
        NotificationCenter.default.addObserver(self, selector: #selector(self.doNothing), name: NSNotification.Name(rawValue: "SomethingChanged"), object: nil)
    }
    
    override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
        
        NotificationCenter.default.removeObserver(self, name: NSNotification.Name(rawValue: "doNothing"), object: nil)
    }
    
    func doNothing() {
        
        // Do nothing
        
        self.loadMeasurements()
        self.htmlView.loadHTMLString(self.createHTML() as String, baseURL: nil)
    }

    @IBAction func shareButtonPressed(_ sender: UIBarButtonItem) {
        
        let alertController = UIAlertController(title: "Share", message: "", preferredStyle: .actionSheet)
        
        let emailAction = UIAlertAction(title: "Email", style: .default, handler: {
            action in
            
            self.emailSummary()
        })
        
        let facebookAction = UIAlertAction(title: "Facebook", style: .default, handler: {
            action in
            
            if(SLComposeViewController.isAvailable(forServiceType: SLServiceTypeFacebook)) {
                let socialController = SLComposeViewController(forServiceType: SLServiceTypeFacebook)
                //            socialController.setInitialText("Hello World!")
                //            socialController.addImage(someUIImageInstance)
                //            socialController.addURL(someNSURLInstance)
                
                self.present(socialController!, animated: true, completion: nil)
            }
            else {
                
                let alertControllerError = UIAlertController(title: "Error", message: "Please ensure you are connected to the internet AND signed into the Facebook app on your device before posting to Facebook.", preferredStyle: .alert)
                
                let cancelActionError = UIAlertAction(title: "OK", style: .cancel, handler: nil)
                
                alertControllerError.addAction(cancelActionError)
                
                if let popoverError = alertControllerError.popoverPresentationController {
                    
                    popoverError.barButtonItem = sender
                    popoverError.sourceView = self.view
                    popoverError.delegate = self
                    popoverError.permittedArrowDirections = .any
                }
                
                self.present(alertControllerError, animated: true, completion: nil)
            }
        })
        
        let twitterAction = UIAlertAction(title: "Twitter", style: .default, handler: {
            action in
            
            if(SLComposeViewController.isAvailable(forServiceType: SLServiceTypeTwitter)) {
                let socialController = SLComposeViewController(forServiceType: SLServiceTypeTwitter)
                //            socialController.setInitialText("Hello World!")
                //            socialController.addImage(someUIImageInstance)
                //            socialController.addURL(someNSURLInstance)
                
                self.present(socialController!, animated: true, completion: nil)
            }
            else {
                
                let alertControllerError = UIAlertController(title: "Error", message: "Please ensure you are connected to the internet AND signed into the Twitter app on your device before posting to Twitter.", preferredStyle: .alert)
                
                let cancelActionError = UIAlertAction(title: "OK", style: .cancel, handler: nil)
                
                alertControllerError.addAction(cancelActionError)
                
                if let popoverError = alertControllerError.popoverPresentationController {
                    
                    popoverError.barButtonItem = sender
                    popoverError.sourceView = self.view
                    popoverError.delegate = self
                    popoverError.permittedArrowDirections = .any
                }
                
                self.present(alertControllerError, animated: true, completion: nil)
            }
        })
        
        let cancelAction = UIAlertAction(title: "Cancel", style: .cancel, handler: nil)
        
        alertController.addAction(emailAction)
        alertController.addAction(facebookAction)
        alertController.addAction(twitterAction)
        alertController.addAction(cancelAction)
        
        if let popover = alertController.popoverPresentationController {
            
            popover.barButtonItem = sender
            popover.sourceView = self.view
            popover.delegate = self
            popover.permittedArrowDirections = .any
        }
        
        present(alertController, animated: true, completion: nil)
    }
    
    func emailSummary() {
        
        // Create MailComposerViewController object.
        let mailcomposer = MFMailComposeViewController()
        mailcomposer.mailComposeDelegate = self
        mailcomposer.navigationBar.tintColor = UIColor.white
        
        let writeString = NSMutableString()
        
        // Check to see if the device has at least 1 email account configured
        if MFMailComposeViewController.canSendMail() {
            
            // Create an array of measurements to iterate thru when building the table rows.
            let measurementsArray = [self.month1Array, self.month2Array, self.month3Array, self.finalArray]
            let measurementsMonth = ["Start Month 1", "Start Month 2", "Start Month 3", "Final"]
            
            writeString.append("Session,Month,Weight,Chest,Left Arm,Right Arm,Waist,Hips,Left Thigh,Right Thigh\n")
            
            for i in 0..<measurementsMonth.count {
                
                writeString.appendFormat("%@,%@", self.session, measurementsMonth[i])
                
                for j in 0..<self.month1Array.count {
                    
                    writeString.appendFormat(",%@", measurementsArray[i][j] as! String)
                }
                
                writeString.append("\n")
            }
        }
        
        // Send email
        let csvData = writeString.data(using: String.Encoding.ascii.rawValue)
        let subject = NSString .localizedStringWithFormat("90 DWT 3 %@ Measurements - Session %@", self.navigationItem.title!, session)
        let fileName = NSString .localizedStringWithFormat("90 DWT 3 %@ Measurements - Session %@.csv", self.navigationItem.title!, session)
        var emailAddress = [""]
        
        // Fetch defaultEmail data.
        let request = NSFetchRequest<NSFetchRequestResult>( entityName: "Email")
        let sortDate = NSSortDescriptor( key: "date", ascending: true)
        request.sortDescriptors = [sortDate]
        
        do {
            if let emailObjects = try CoreDataHelper.shared().context.fetch(request) as? [Email] {
                
                if debug == 1 {
                    
                    print("emailObjects.count = \(emailObjects.count)")
                }
                
                if emailObjects.count != 0 {
                    
                    // There is a default email address.
                    emailAddress = [(emailObjects.last?.defaultEmail)!]
                }
                else {
                    
                    // There is NOT a default email address.  Put an empty email address in the arrary.
                    emailAddress = [""]
                }
            }
            
        } catch { print(" ERROR executing a fetch request: \( error)") }
        
        mailcomposer.setToRecipients(emailAddress)
        mailcomposer.setSubject(subject as String)
        mailcomposer.addAttachmentData(csvData!, mimeType: "text/csv", fileName: fileName as String)
        
        present(mailcomposer, animated: true, completion: {
            UIApplication.shared.statusBarStyle = UIStatusBarStyle.lightContent
        })
    }
    
    func mailComposeController(_ controller: MFMailComposeViewController, didFinishWith result: MFMailComposeResult, error: Error?) {
        
        dismiss(animated: true, completion: nil)
    }

    func loadMeasurements() {
        
        let monthArray = ["1",
                          "2",
                          "3",
                          "4"]
        
        // Get workout data with the current session
        let request = NSFetchRequest<NSFetchRequestResult>( entityName: "Measurement")
        let sortDate = NSSortDescriptor( key: "date", ascending: true)
        request.sortDescriptors = [sortDate]
        
        for i in 0..<monthArray.count {
            
            let filter = NSPredicate(format: "session == %@ AND month == %@",
                                     session,
                                     monthArray[i])
            
            request.predicate = filter

            do {
                if let measurementObjects = try CoreDataHelper.shared().context.fetch(request) as? [Measurement] {
                    
                    if debug == 1 {
                        
                        print("measurementObjects.count = \(measurementObjects.count)")
                    }
                    
                    if measurementObjects.count >= 1 {
                        
                        if monthArray[i] == "1" {
                            
                            // Weight
                            if measurementObjects.last?.weight == "" || measurementObjects.last?.weight == nil {
                                self.month1Array.add("0")
                            }
                            else {
                                self.month1Array.add((measurementObjects.last?.weight)!)
                            }
                            
                            // Chest
                            if measurementObjects.last?.chest == "" || measurementObjects.last?.chest == nil {
                                self.month1Array.add("0")
                            }
                            else {
                                self.month1Array.add((measurementObjects.last?.chest)!)
                            }
                            
                            // Left Arm
                            if measurementObjects.last?.leftArm == "" || measurementObjects.last?.leftArm == nil {
                                self.month1Array.add("0")
                            }
                            else {
                                self.month1Array.add((measurementObjects.last?.leftArm)!)
                            }
                            
                            // Right Arm
                            if measurementObjects.last?.rightArm == "" || measurementObjects.last?.rightArm == nil {
                                self.month1Array.add("0")
                            }
                            else {
                                self.month1Array.add((measurementObjects.last?.rightArm)!)
                            }
                            
                            // Waist
                            if measurementObjects.last?.waist == "" || measurementObjects.last?.waist == nil {
                                self.month1Array.add("0")
                            }
                            else {
                                self.month1Array.add((measurementObjects.last?.waist)!)
                            }
                            
                            // Hips
                            if measurementObjects.last?.hips == "" || measurementObjects.last?.hips == nil {
                                self.month1Array.add("0")
                            }
                            else {
                                self.month1Array.add((measurementObjects.last?.hips)!)
                            }
                            
                            // Left Thigh
                            if measurementObjects.last?.leftThigh == "" || measurementObjects.last?.leftThigh == nil {
                                self.month1Array.add("0")
                            }
                            else {
                                self.month1Array.add((measurementObjects.last?.leftThigh)!)
                            }
                            
                            // Right Thigh
                            if measurementObjects.last?.rightThigh == "" || measurementObjects.last?.rightThigh == nil {
                                self.month1Array.add("0")
                            }
                            else {
                                self.month1Array.add((measurementObjects.last?.rightThigh)!)
                            }
                        }
                        
                        else if monthArray[i] == "2" {
                            
                            // Weight
                            if measurementObjects.last?.weight == "" || measurementObjects.last?.weight == nil {
                                self.month2Array.add("0")
                            }
                            else {
                                self.month2Array.add((measurementObjects.last?.weight)!)
                            }
                            
                            // Chest
                            if measurementObjects.last?.chest == "" || measurementObjects.last?.chest == nil {
                                self.month2Array.add("0")
                            }
                            else {
                                self.month2Array.add((measurementObjects.last?.chest)!)
                            }
                            
                            // Left Arm
                            if measurementObjects.last?.leftArm == "" || measurementObjects.last?.leftArm == nil {
                                self.month2Array.add("0")
                            }
                            else {
                                self.month2Array.add((measurementObjects.last?.leftArm)!)
                            }
                            
                            // Right Arm
                            if measurementObjects.last?.rightArm == "" || measurementObjects.last?.rightArm == nil {
                                self.month2Array.add("0")
                            }
                            else {
                                self.month2Array.add((measurementObjects.last?.rightArm)!)
                            }
                            
                            // Waist
                            if measurementObjects.last?.waist == "" || measurementObjects.last?.waist == nil {
                                self.month2Array.add("0")
                            }
                            else {
                                self.month2Array.add((measurementObjects.last?.waist)!)
                            }
                            
                            // Hips
                            if measurementObjects.last?.hips == "" || measurementObjects.last?.hips == nil {
                                self.month2Array.add("0")
                            }
                            else {
                                self.month2Array.add((measurementObjects.last?.hips)!)
                            }
                            
                            // Left Thigh
                            if measurementObjects.last?.leftThigh == "" || measurementObjects.last?.leftThigh == nil {
                                self.month2Array.add("0")
                            }
                            else {
                                self.month2Array.add((measurementObjects.last?.leftThigh)!)
                            }
                            
                            // Right Thigh
                            if measurementObjects.last?.rightThigh == "" || measurementObjects.last?.rightThigh == nil {
                                self.month2Array.add("0")
                            }
                            else {
                                self.month2Array.add((measurementObjects.last?.rightThigh)!)
                            }
                        }

                        else if monthArray[i] == "3" {
                            
                            // Weight
                            if measurementObjects.last?.weight == "" || measurementObjects.last?.weight == nil {
                                self.month3Array.add("0")
                            }
                            else {
                                self.month3Array.add((measurementObjects.last?.weight)!)
                            }
                            
                            // Chest
                            if measurementObjects.last?.chest == "" || measurementObjects.last?.chest == nil {
                                self.month3Array.add("0")
                            }
                            else {
                                self.month3Array.add((measurementObjects.last?.chest)!)
                            }
                            
                            // Left Arm
                            if measurementObjects.last?.leftArm == "" || measurementObjects.last?.leftArm == nil {
                                self.month3Array.add("0")
                            }
                            else {
                                self.month3Array.add((measurementObjects.last?.leftArm)!)
                            }
                            
                            // Right Arm
                            if measurementObjects.last?.rightArm == "" || measurementObjects.last?.rightArm == nil {
                                self.month3Array.add("0")
                            }
                            else {
                                self.month3Array.add((measurementObjects.last?.rightArm)!)
                            }
                            
                            // Waist
                            if measurementObjects.last?.waist == "" || measurementObjects.last?.waist == nil {
                                self.month3Array.add("0")
                            }
                            else {
                                self.month3Array.add((measurementObjects.last?.waist)!)
                            }
                            
                            // Hips
                            if measurementObjects.last?.hips == "" || measurementObjects.last?.hips == nil {
                                self.month3Array.add("0")
                            }
                            else {
                                self.month3Array.add((measurementObjects.last?.hips)!)
                            }
                            
                            // Left Thigh
                            if measurementObjects.last?.leftThigh == "" || measurementObjects.last?.leftThigh == nil {
                                self.month3Array.add("0")
                            }
                            else {
                                self.month3Array.add((measurementObjects.last?.leftThigh)!)
                            }
                            
                            // Right Thigh
                            if measurementObjects.last?.rightThigh == "" || measurementObjects.last?.rightThigh == nil {
                                self.month3Array.add("0")
                            }
                            else {
                                self.month3Array.add((measurementObjects.last?.rightThigh)!)
                            }
                        }

                        else if monthArray[i] == "4" {
                            
                            // Weight
                            if measurementObjects.last?.weight == "" || measurementObjects.last?.weight == nil {
                                self.finalArray.add("0")
                            }
                            else {
                                self.finalArray.add((measurementObjects.last?.weight)!)
                            }
                            
                            // Chest
                            if measurementObjects.last?.chest == "" || measurementObjects.last?.chest == nil {
                                self.finalArray.add("0")
                            }
                            else {
                                self.finalArray.add((measurementObjects.last?.chest)!)
                            }
                            
                            // Left Arm
                            if measurementObjects.last?.leftArm == "" || measurementObjects.last?.leftArm == nil {
                                self.finalArray.add("0")
                            }
                            else {
                                self.finalArray.add((measurementObjects.last?.leftArm)!)
                            }
                            
                            // Right Arm
                            if measurementObjects.last?.rightArm == "" || measurementObjects.last?.rightArm == nil {
                                self.finalArray.add("0")
                            }
                            else {
                                self.finalArray.add((measurementObjects.last?.rightArm)!)
                            }
                            
                            // Waist
                            if measurementObjects.last?.waist == "" || measurementObjects.last?.waist == nil {
                                self.finalArray.add("0")
                            }
                            else {
                                self.finalArray.add((measurementObjects.last?.waist)!)
                            }
                            
                            // Hips
                            if measurementObjects.last?.hips == "" || measurementObjects.last?.hips == nil {
                                self.finalArray.add("0")
                            }
                            else {
                                self.finalArray.add((measurementObjects.last?.hips)!)
                            }
                            
                            // Left Thigh
                            if measurementObjects.last?.leftThigh == "" || measurementObjects.last?.leftThigh == nil {
                                self.finalArray.add("0")
                            }
                            else {
                                self.finalArray.add((measurementObjects.last?.leftThigh)!)
                            }
                            
                            // Right Thigh
                            if measurementObjects.last?.rightThigh == "" || measurementObjects.last?.rightThigh == nil {
                                self.finalArray.add("0")
                            }
                            else {
                                self.finalArray.add((measurementObjects.last?.rightThigh)!)
                            }
                        }
                    }
                    
                    else {
                        
                        if monthArray[i] == "1" {
                            
                            self.month1Array.add("0")
                            self.month1Array.add("0")
                            self.month1Array.add("0")
                            self.month1Array.add("0")
                            self.month1Array.add("0")
                            self.month1Array.add("0")
                            self.month1Array.add("0")
                            self.month1Array.add("0")
                        }
                        
                        else if monthArray[i] == "2" {
                            
                            self.month2Array.add("0")
                            self.month2Array.add("0")
                            self.month2Array.add("0")
                            self.month2Array.add("0")
                            self.month2Array.add("0")
                            self.month2Array.add("0")
                            self.month2Array.add("0")
                            self.month2Array.add("0")
                        }
                        
                        else if monthArray[i] == "3" {
                            
                            self.month3Array.add("0")
                            self.month3Array.add("0")
                            self.month3Array.add("0")
                            self.month3Array.add("0")
                            self.month3Array.add("0")
                            self.month3Array.add("0")
                            self.month3Array.add("0")
                            self.month3Array.add("0")
                        }
                        
                        else if monthArray[i] == "4" {
                            
                            self.finalArray.add("0")
                            self.finalArray.add("0")
                            self.finalArray.add("0")
                            self.finalArray.add("0")
                            self.finalArray.add("0")
                            self.finalArray.add("0")
                            self.finalArray.add("0")
                            self.finalArray.add("0")
                        }
                    }
                }
                
            } catch { print(" ERROR executing a fetch request: \( error)") }
        }
    }
    
    func createHTML() -> NSString {
        
        // Create an array of measurements to iterate thru when building the table rows.
        let measurementsArray = [self.month1Array, self.month2Array, self.month3Array, self.finalArray]
        let measurementsNameArray = ["Weight", "Chest", "Left Arm", "Right Arm", "Waist", "Hips", "Left Thigh", "Right Thigh"]
        
        var myHTML = "<html><head>"
        
        // Table Style
        myHTML = myHTML.appendingFormat("<STYLE TYPE='text/css'><!--TD{font-family: Arial; font-size: 12pt;}TH{font-family: Arial; font-size: 14pt;}---></STYLE></head><body><table border='1' bordercolor='#3399FF' style='background-color:#CCCCCC' width='%f' cellpadding='2' cellspacing='1'>", self.htmlView.frame.size.width - 15)
        
        // Table Headers
        myHTML = myHTML.appendingFormat("<tr><th style='background-color:#999999'>Session %@</th><th style='background-color:#999999'>1</th><th style='background-color:#999999'>2</th><th style='background-color:#999999'>3</th><th style='background-color:#999999'>Final</th></tr>", session)
        
        // Table Data
        for i in 0..<measurementsNameArray.count {
            
            myHTML = myHTML.appendingFormat("<tr><td style='background-color:#999999'>%@</td>", measurementsNameArray[i])
            
            for a in 0..<measurementsArray.count {
                
                myHTML = myHTML.appendingFormat("<td>%@</td>", measurementsArray[a][i] as! String)
            }
            
            myHTML = myHTML + "</tr>"
        }
        
        // HTML closing tags
        myHTML = myHTML + "</table></body></html>"
        
        return myHTML as NSString
    }
}
